src/game.cpp

```diff
@@ -4919,6 +4921,92 @@ void Game::playerDebugAssert(uint32_t playerId, const std::string& assertLine, c
+	Player* player = getPlayerByID(playerId);
+	if (!player) {
+		return;
+	}
+
+	auto offer = g_store->getOfferById(offerId);
+	if (!offer) {
+		player->sendStoreError(STORE_ERROR_PURCHASE, "Offer not found. Please reopen the store and try again.");
+		return;
+	}
+
+	auto coinBalance = IOAccount::getCoinBalance(player->getAccount());
+	if (coinBalance >= offer->getPrice()) {
+		if (g_store->executeOnBuy(player, &(*offer), param)) {
+			IOAccount::addCoins(player->getAccount(), -static_cast<int32_t>(offer->getPrice()));
+			player->coinBalance -= offer->getPrice();
+			player->sendStorePurchaseCompleted(offer->getMessage());
+			player->sendTextMessage(MESSAGE_EVENT_ADVANCE, "A transaction was successful. For more details, please check your Store Coins history.");
+
+			std::string transactionDescription;
+			if (offer->getSubOffers().size() == 0) {
+				transactionDescription = offer->getName();
+			} else {
+				std::ostringstream tmpDesc;
+				bool first = true;
+				for (const auto& subOffer : offer->getSubOffers()) {
+					if (!first) {
+						tmpDesc << ", ";
+					} else {
+						first = false;
+					}
+					tmpDesc << subOffer.getName();
+				}
+				transactionDescription = tmpDesc.str();
+			}
+
+			g_store->onTransactionCompleted(player->getAccount(), -static_cast<int32_t>(offer->getPrice()), transactionDescription);
+		}
+	} else {
+		player->sendStoreError(STORE_ERROR_PURCHASE, "You do not have enough coins to complete this purchase.");
+	}
+}
+
+void Game::playerTransferCoins(uint32_t playerId, const std::string& recipient, uint16_t amount) {
+	Player* player = getPlayerByID(playerId);
+	if (!player) {
+		return;
+	}
+
+	if (player->coinBalance < amount) {
+		return player->sendStoreError(STORE_ERROR_TRANSFER, "You don't have enough coins to transfer.");
+	}
+
+	if (amount % g_config.getNumber(ConfigManager::STORE_COINS_PACKET_SIZE) != 0) {
+		return player->sendStoreError(STORE_ERROR_TRANSFER, "You can't transfer this amount of coins.");
+	}
+
+	Player *tmpPlayer = g_game.getPlayerByName(recipient);
+	if (!tmpPlayer) {
+		tmpPlayer = new Player(nullptr);
+		if (!IOLoginData::loadPlayerByName(tmpPlayer, recipient)) {
+			delete tmpPlayer;
+			return player->sendStoreError(STORE_ERROR_TRANSFER, "Recipient player not found");
+		}
+	}
+
+	if (tmpPlayer->getAccount() == player->getAccount()) {
+		return player->sendStoreError(STORE_ERROR_TRANSFER, "You can't transfer coins to the same account.");
+	}
+
+	std::string description(player->getName() + " transferred to " + recipient);
+
+	IOAccount::addCoins(player->getAccount(), -static_cast<int32_t>(amount));
+	g_store->onTransactionCompleted(player->getAccount(), -static_cast<int32_t>(amount), description);
+
+	IOAccount::addCoins(tmpPlayer->getAccount(), amount);
+	g_store->onTransactionCompleted(tmpPlayer->getAccount(), amount, description);
+
+	if (tmpPlayer->isOffline()) {
+		delete tmpPlayer;
```

Comment:
Just saying, saw the game market code after reading the ninja comment and when you receive something on market if you offline then you will be saved.